presto-common/src/main/java/com/facebook/presto/common/block/MapBlockBuilder.java

```diff
@@ -0,0 +1,635 @@
+                keyBlockBuilder.newBlockBuilderLike(blockBuilderStatus, nestedExpectedEntries),
+                valueBlockBuilder.newBlockBuilderLike(blockBuilderStatus, nestedExpectedEntries),
+                newSize,
+                getNewHashTables(newSize));
+    }
+
+    public void loadHashTables()
+    {
+        ensureHashTableLoaded(keyBlockHashCode);
+    }
+
+    @Override
+    protected void ensureHashTableLoaded(MethodHandle keyBlockHashCode)
+    {
+        // MapBlockBuilder does not support concurrent threads, so synchronization is not required.
+        if (!isHashTablesPresent()) {
+            hashTables.loadHashTables(positionCount, offsets, mapIsNull, keyBlockBuilder, keyBlockHashCode);
+        }
+    }
+
+    private static int[] newNegativeOneFilledArray(int size)
+    {
+        int[] hashTable = new int[size];
+        Arrays.fill(hashTable, -1);
+        return hashTable;
+    }
+
+    /**
+     * This method assumes that {@code keyBlock} has no duplicated entries (in the specified range)
+     */
+    static void buildHashTable(Block keyBlock, int keyOffset, int keyCount, MethodHandle keyBlockHashCode, int[] outputHashTable, int hashTableOffset, int hashTableSize)
+    {
+        for (int i = 0; i < keyCount; i++) {
+            int hash = getHashPosition(keyBlock, keyOffset + i, keyBlockHashCode, hashTableSize);
+            while (true) {
+                if (outputHashTable[hashTableOffset + hash] == -1) {
+                    outputHashTable[hashTableOffset + hash] = i;
+                    break;
+                }
+                hash++;
+                if (hash == hashTableSize) {
+                    hash = 0;
+                }
+            }
+        }
+    }
+
+    /**
+     * This method checks whether {@code keyBlock} has duplicated entries (in the specified range)
+     */
+    private static void buildHashTableStrict(
+            Block keyBlock,
+            int keyOffset,
+            int keyCount,
+            MethodHandle keyBlockEquals,
+            MethodHandle keyBlockHashCode,
+            int[] outputHashTable,
+            int hashTableOffset,
+            int hashTableSize)
+            throws DuplicateMapKeyException, NotSupportedException
+    {
+        for (int i = 0; i < keyCount; i++) {
+            int hash = getHashPosition(keyBlock, keyOffset + i, keyBlockHashCode, hashTableSize);
+            while (true) {
+                if (outputHashTable[hashTableOffset + hash] == -1) {
+                    outputHashTable[hashTableOffset + hash] = i;
+                    break;
+                }
+
+                Boolean isDuplicateKey;
+                try {
+                    // assuming maps with indeterminate keys are not supported
+                    isDuplicateKey = (Boolean) keyBlockEquals.invokeExact(keyBlock, keyOffset + i, keyBlock, keyOffset + outputHashTable[hashTableOffset + hash]);
+                }
+                catch (RuntimeException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: An exception is caught and re-thrown within the catch clause without any other operations being performed on the exception, making the catch clause unnecessary.

**Fix**: Perform additional operations on the exception that has been caught. You can wrap a standard exception into a custom exception to add additional information or special handling. When you do that, make sure to set the original exception as the cause. Otherwise, you will lose the stack trace and message of the original exception, which will make it difficult to analyze the event that caused the exception.  
[Learn more](https://www.ibm.com/support/pages/best-practice-catching-and-re-throwing-java-exceptions)